File: mdtomd/config/simple_yaml.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class _YamlLine:
    indent: int
    content: str
    line_no: int


def load_simple_yaml(text: str) -> dict[str, Any]:
    lines = _prepare_yaml_lines(text)
    if not lines:
        return {}
    value, index = _parse_yaml_block(lines, 0, lines[0].indent)
    if index != len(lines):
        raise ValueError("Unexpected trailing content in config.yaml")
    if not isinstance(value, dict):
        raise ValueError("Config root must be a mapping")
    return value


def _prepare_yaml_lines(text: str) -> list[_YamlLine]:
    prepared: list[_YamlLine] = []
    for line_no, raw_line in enumerate(text.splitlines(), start=1):
        stripped = _strip_inline_comment(raw_line).rstrip()
        if not stripped.strip():
            continue
        if "\t" in raw_line:
            raise ValueError(f"Tabs are not supported in config.yaml (line {line_no})")
        indent = len(stripped) - len(stripped.lstrip(" "))
        prepared.append(
            _YamlLine(
                indent=indent,
                content=stripped.strip(),
                line_no=line_no,
            )
        )
    return prepared


def _strip_inline_comment(line: str) -> str:
    in_single = False
    in_double = False
    for index, char in enumerate(line):
        if char == "'" and not in_double:
            in_single = not in_single
        elif char == '"' and not in_single:
            in_double = not in_double
        elif char == "#" and not in_single and not in_double:
            if index == 0 or line[index - 1].isspace():
                return line[:index]
    return line


def _parse_yaml_block(lines: list[_YamlLine], index: int, indent: int) -> tuple[Any, int]:
    if index >= len(lines):
        return {}, index
    if lines[index].content.startswith("- "):
        return _parse_yaml_sequence(lines, index, indent)
    return _parse_yaml_mapping(lines, index, indent)


def _parse_yaml_mapping(lines: list[_YamlLine], index: int, indent: int) -> tuple[dict[str, Any], int]:
    result: dict[str, Any] = {}
    while index < len(lines):
        line = lines[index]
        if line.indent < indent:
            break
        if line.indent > indent:
            raise ValueError(f"Invalid indentation at line {line.line_no}")
        if line.content.startswith("- "):
            break

        key, raw_value = _split_key_value(line.content, line.line_no)
        index += 1
        if raw_value == "":
            if index < len(lines) and lines[index].indent > indent:
                value, index = _parse_yaml_block(lines, index, lines[index].indent)
            else:
                value = {}
        else:
            value = _parse_scalar(raw_value)
        result[key] = value
    return result, index


def _parse_yaml_sequence(lines: list[_YamlLine], index: int, indent: int) -> tuple[list[Any], int]:
    result: list[Any] = []
    while index < len(lines):
        line = lines[index]
        if line.indent < indent:
            break
        if line.indent > indent:
            raise ValueError(f"Invalid indentation at line {line.line_no}")
        if not line.content.startswith("- "):
            break

        item_text = line.content[2:].strip()
        index += 1
        if item_text == "":
            if index < len(lines) and lines[index].indent > indent:
                value, index = _parse_yaml_block(lines, index, lines[index].indent)
            else:
                value = None
            result.append(value)
            continue

        if ":" in item_text and not item_text.startswith(("http://", "https://")):
            key, raw_value = _split_key_value(item_text, line.line_no)
            if raw_value == "":
                if index < len(lines) and lines[index].indent > indent:
                    value, index = _parse_yaml_block(lines, index, lines[index].indent)
                else:
                    value = {}
            else:
                value = _parse_scalar(raw_value)
            result.append({key: value})
            continue

        result.append(_parse_scalar(item_text))
    return result, index


def _split_key_value(content: str, line_no: int) -> tuple[str, str]:
    if ":" not in content:
        raise ValueError(f"Expected key: value at line {line_no}")
    key, raw_value = content.split(":", 1)
    normalized_key = key.strip()
    if not normalized_key:
        raise ValueError(f"Empty key at line {line_no}")
    return normalized_key, raw_value.strip()


def _parse_scalar(value: str) -> Any:
    normalized = value.strip()
    if not normalized:
        return ""
    if normalized[0] == normalized[-1] and normalized[0] in {"'", '"'}:
        return normalized[1:-1]

    lowered = normalized.lower()
    if lowered in {"true", "yes", "on"}:
        return True
    if lowered in {"false", "no", "off"}:
        return False
    if lowered in {"null", "none", "~"}:
        return None

    try:
        if normalized.startswith("0") and normalized not in {"0", "0.0"} and not normalized.startswith("0."):
            raise ValueError
        return int(normalized)
    except ValueError:
        pass

    try:
        return float(normalized)
    except ValueError:
        return normalized

File: vscode-extension/lib/simpleYaml.js
function loadSimpleYaml(text) {
  const lines = prepareYamlLines(text);
  if (!lines.length) {
    return {};
  }

  const [value, index] = parseYamlBlock(lines, 0, lines[0].indent);
  if (index !== lines.length) {
    throw new Error('Unexpected trailing content in config.yaml');
  }
  if (!isPlainObject(value)) {
    throw new Error('Config root must be a mapping');
  }
  return value;
}

function prepareYamlLines(text) {
  const prepared = [];
  const rawLines = String(text || '').split(/\r?\n/u);
  for (let index = 0; index < rawLines.length; index += 1) {
    const lineNumber = index + 1;
    const rawLine = rawLines[index];
    const stripped = stripInlineComment(rawLine).replace(/\s+$/u, '');
    if (!stripped.trim()) {
      continue;
    }
    if (rawLine.includes('\t')) {
      throw new Error(`Tabs are not supported in config.yaml (line ${lineNumber})`);
    }

    const indent = stripped.length - stripped.replace(/^ */u, '').length;
    prepared.push({
      indent,
      content: stripped.trim(),
      lineNumber,
    });
  }
  return prepared;
}

function stripInlineComment(line) {
  let inSingle = false;
  let inDouble = false;
  for (let index = 0; index < line.length; index += 1) {
    const char = line[index];
    if (char === "'" && !inDouble) {
      inSingle = !inSingle;
      continue;
    }
    if (char === '"' && !inSingle) {
      inDouble = !inDouble;
      continue;
    }
    if (char === '#' && !inSingle && !inDouble) {
      if (index === 0 || /\s/u.test(line[index - 1])) {
        return line.slice(0, index);
      }
    }
  }
  return line;
}

function parseYamlBlock(lines, index, indent) {
  if (index >= lines.length) {
    return [{}, index];
  }
  if (lines[index].content.startsWith('- ')) {
    return parseYamlSequence(lines, index, indent);
  }
  return parseYamlMapping(lines, index, indent);
}

function parseYamlMapping(lines, index, indent) {
  const result = {};
  let currentIndex = index;

  while (currentIndex < lines.length) {
    const line = lines[currentIndex];
    if (line.indent < indent) {
      break;
    }
    if (line.indent > indent) {
      throw new Error(`Invalid indentation at line ${line.lineNumber}`);
    }
    if (line.content.startsWith('- ')) {
      break;
    }

    const [key, rawValue] = splitKeyValue(line.content, line.lineNumber);
    currentIndex += 1;
    let value;
    if (rawValue === '') {
      if (currentIndex < lines.length && lines[currentIndex].indent > indent) {
        [value, currentIndex] = parseYamlBlock(lines, currentIndex, lines[currentIndex].indent);
      } else {
        value = {};
      }
    } else {
      value = parseScalar(rawValue);
    }
    result[key] = value;
  }

  return [result, currentIndex];
}

function parseYamlSequence(lines, index, indent) {
  const result = [];
  let currentIndex = index;

  while (currentIndex < lines.length) {
    const line = lines[currentIndex];
    if (line.indent < indent) {
      break;
    }
    if (line.indent > indent) {
      throw new Error(`Invalid indentation at line ${line.lineNumber}`);
    }
    if (!line.content.startsWith('- ')) {
      break;
    }

    const itemText = line.content.slice(2).trim();
    currentIndex += 1;
    if (itemText === '') {
      let nestedValue = null;
      if (currentIndex < lines.length && lines[currentIndex].indent > indent) {
        [nestedValue, currentIndex] = parseYamlBlock(lines, currentIndex, lines[currentIndex].indent);
      }
      result.push(nestedValue);
      continue;
    }

    if (itemText.includes(':') && !itemText.startsWith('http://') && !itemText.startsWith('https://')) {
      const [key, rawValue] = splitKeyValue(itemText, line.lineNumber);
      let value;
      if (rawValue === '') {
        if (currentIndex < lines.length && lines[currentIndex].indent > indent) {
          [value, currentIndex] = parseYamlBlock(lines, currentIndex, lines[currentIndex].indent);
        } else {
          value = {};
        }
      } else {
        value = parseScalar(rawValue);
      }
      result.push({ [key]: value });
      continue;
    }

    result.push(parseScalar(itemText));
  }

  return [result, currentIndex];
}

function splitKeyValue(content, lineNumber) {
  const separatorIndex = content.indexOf(':');
  if (separatorIndex === -1) {
    throw new Error(`Expected key: value at line ${lineNumber}`);
  }
  const key = content.slice(0, separatorIndex).trim();
  if (!key) {
    throw new Error(`Empty key at line ${lineNumber}`);
  }
  return [key, content.slice(separatorIndex + 1).trim()];
}

function parseScalar(value) {
  const normalized = String(value || '').trim();
  if (!normalized) {
    return '';
  }
  if (normalized.length >= 2 && normalized[0] === normalized[normalized.length - 1] && (normalized[0] === "'" || normalized[0] === '"')) {
    return normalized.slice(1, -1);
  }

  const lowered = normalized.toLowerCase();
  if (['true', 'yes', 'on'].includes(lowered)) {
    return true;
  }
  if (['false', 'no', 'off'].includes(lowered)) {
    return false;
  }
  if (['null', 'none', '~'].includes(lowered)) {
    return null;
  }

  if (shouldParseInt(normalized)) {
    return Number.parseInt(normalized, 10);
  }
  if (/^-?\d+\.\d+$/u.test(normalized)) {
    return Number.parseFloat(normalized);
  }
  return normalized;
}

function shouldParseInt(value) {
  if (!/^-?\d+$/u.test(value)) {
    return false;
  }
  if (value === '0' || value === '-0') {
    return true;
  }
  const unsigned = value.startsWith('-') ? value.slice(1) : value;
  return !unsigned.startsWith('0');
}

function isPlainObject(value) {
  return value && typeof value === 'object' && !Array.isArray(value);
}

module.exports = {
  loadSimpleYaml,
};
